http-client/src/main/java/io/airlift/http/client/jetty/HttpRequestEvent.java

```diff
@@ -0,0 +1,177 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.http.client.jetty;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.http.client.jetty.HttpClientLogger.RequestInfo;
+import io.airlift.http.client.jetty.HttpClientLogger.ResponseInfo;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.api.Response;
+import org.eclipse.jetty.http.HttpFields;
+
+import javax.annotation.Nullable;
+
+import java.time.Instant;
+import java.util.Optional;
+
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+class HttpRequestEvent
+{
+    static final int NO_RESPONSE = -1;
+
+    private final Instant timeStamp;
+    private final String protocol;
+    private final String method;
+    private final String requestUri;
+
+    // Response-related fields can have the value of NO_RESPONSE for the requests
+    // that didn't get any response (e.g., due to timeouts or other failures).
+    private final long responseSize;
+    private final int responseCode;
+    private final long timeToLastByte;
+    private final Optional<String> failureReason;
+
+    private HttpRequestEvent(
+            Instant timeStamp,
+            String protocol,
+            String method,
+            String requestUri,
+            long responseSize,
+            int responseCode,
+            long timeToLastByte,
+            Optional<String> failureReason)
+    {
+        this.timeStamp = requireNonNull(timeStamp, "timeStamp is null");
+        this.protocol = requireNonNull(protocol, "protocol is null");
+        this.method = requireNonNull(method, "method is null");
+        this.requestUri = requireNonNull(requestUri, "requestUri is null");
+        this.responseSize = responseSize;
+        this.responseCode = responseCode;
+        this.timeToLastByte = timeToLastByte;
+        this.failureReason = requireNonNull(failureReason, "failureReason is null");
+    }
+
+    public Instant getTimeStamp()
+    {
+        return timeStamp;
+    }
+
+    public String getProtocol()
+    {
+        return protocol;
+    }
+
+    public String getMethod()
+    {
+        return method;
+    }
+
+    public String getRequestUri()
+    {
+        return requestUri;
+    }
+
+    public long getResponseSize()
+    {
+        return responseSize;
+    }
+
+    public String getResponseCode()
+    {
+        return failureReason.orElseGet(() -> Integer.toString(responseCode));
```

Comment:
I find this strange.  It means the response code column in the log can be a non numeric value.

Reply:
Yes. For example, for requests that fail with a timeout exception that column will be `TIMEOUT` as we don't have a response code. What do we want to have in the logs for such failed requests? Maybe we can have both response code and failure reason columns, and for cases like these former can be empty, etc.

Reply:
Keeping it as is as we discussed offline.